### PR TITLE
Lfa1253/sign on page

### PR DESCRIPTION
**Pull request by franciscog-kainos**
### JIRA link
https://companieshouse.atlassian.net/browse/LFA-1260

### Change description
Added `ch-session-node` api which is currently under review [PR for poc-session-handler](https://github.com/companieshouse/poc-node-session-handler/pull/2) 

### Work checklist

- [ ] Tests added where applicable
- [ ] UI changes look good on mobile
- [ ] UI changes meet accessibility criteria

### Merge instructions

We are commited to keeping commit history clean, consistent and linear. To achive this commit should be structured as follows:

```
<type>[optional scope]: <description>
```

and contain the following structural elements:

- fix: a commit that patches a bug in your codebase (this correlates with PATCH in semantic versioning),
- feat: a commit that introduces a new feature to the codebase (this correlates with MINOR in semantic versioning),
- BREAKING CHANGE: a commit that has a footer `BREAKING CHANGE:` introduces a breaking API change (correlating with MAJOR in semantic versioning). A BREAKING CHANGE can be part of commits of any type,
- types other than `fix:` and `feat:` are allowed, for example `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`, and others,
- footers other than `BREAKING CHANGE: <description>` may be provided.

### CODE REVIEW

File: src/views/landing.njk [hunk 1]
```diff
@@ -32,7 +32,8 @@
             isStartButton: true,
             attributes: {
               id: "submit"
-            }
+            },
+            href: penaltyDetailsPage
```
Comment (akaJSaunders) [inline src/views/landing.njk]: I think we were avoiding adding transitions to other pages through the template files. Currently in master, landing page redirects through a post request (when button is clicked) to a entry controller ('/start') which calls a get request that redirects to penalty details page. This is because landing page is ultimately a throwaway page. Is this explanation correct @geek-soft ? haha

File: src/App.ts [hunk 2]
```diff
@@ -1,7 +1,7 @@
 import { Server } from './Server';
-import { loadEnvironmentVariablesFromFiles } from './utils/ConfigLoader';
+import { loadEnvironmentVariablesFromFiles, ENV } from './utils/ConfigLoader';
 
 loadEnvironmentVariablesFromFiles();
 
-const server = new Server(Number(process.env.PORT) || 3000);
+const server = new Server(ENV.ERIC_PORT);
```
Comment (geek-soft) [inline src/App.ts]: How is that related to the user story?

File: src/views/landing.njk (hunk 1 above)
Comment (geek-soft) [reply src/views/landing.njk]: Redirect is done on submission, you are right @akaJSaunders it is not needed.

File: src/utils/IConfig.ts [hunk 3]
```diff
@@ -0,0 +1,10 @@
+export default interface IConfig {
+    readonly ERIC_PORT: number;
+    readonly CACHE_SERVER: string;
+    readonly CACHE_DB: number;
+    readonly CACHE_PASSWORD: string;
+    readonly COOKIE_NAME: string;
+    readonly COOKIE_SECRET: string;
+    readonly DEFAULT_SESSION_EXPIRATION: number;
```
Comment (geek-soft) [inline src/utils/IConfig.ts]: How is that used or intended to be used?

File: src/utils/ConfigLoader.ts [hunk 4]
```diff
@@ -14,11 +15,14 @@ const checkFileExists = (config: dotenv.DotenvConfigOutput) => {
...
 };
 
+export let ENV: IConfig;
+
 export const loadEnvironmentVariablesFromFiles = () => {
   dotenv.config({ path: DEFAULT_ENV_FILE });
   if (process.env.NODE_ENV) {
     const envFilePath = `${__dirname}/../../.env.${process.env.NODE_ENV}`;
     checkFileExists(dotenv.config({ path: envFilePath }));
+    ENV = new Config();
```
Comment (geek-soft) [inline src/utils/ConfigLoader.ts]: It looks like surprising side effect. I guess implementation might not be complete but at this point doing validation in constructor does not look like the right thing.

File: package.json [hunk 5]
```diff
@@ -29,6 +30,7 @@
     "inversify": "^5.0.1",
     "inversify-binding-decorators": "^4.0.0",
     "inversify-express-utils": "^6.3.2",
+    "msgpack5": "^4.2.1",
```
Comment (geek-soft) [inline package.json]: Removal candidate, `ch-node-session` should bring it as transitive dep.

File: src/utils/IConfig.ts (hunk 3 above)
Comment (franciscog-kainos) [reply src/utils/IConfig.ts]: the process.env give out variables of type `T  | undefined`. The point was centralise the checking of ENV variables so that we don't have to check every time if process.env variable is undefined

File: src/middleware/AuthMiddleware.ts [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+export class AuthMiddleware extends BaseMiddleware {
+
+    public constructor() {
+        super();
+    }
+
+    public handler: RequestHandler = expressAsyncHandler((req: Request, res: Response, next: NextFunction): void => {
+
+        if (req.session.isNothing()) {
+            res.redirect(`http://localhost:3001/signin?return_to=${req.url}`);
```
Comment (n-r-b) [inline src/middleware/AuthMiddleware.ts]: Can we create a constant for this path and make the env configurable?

File: src/middleware/AuthMiddleware.ts [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+        }
+        console.log('Authenticated');
```
Comment (n-r-b) [inline src/middleware/AuthMiddleware.ts]: How do we know at this point the sign in was successful?  Or do we assume it was and then have a check in the controller?

Comment (franciscog-kainos) [reply src/middleware/AuthMiddleware.ts]: Check new version

File: src/controllers/OtherReasonController.ts [hunk 8]
```diff
@@ -5,33 +5,49 @@ import { SchemaValidator } from '../utils/validation/SchemaValidator';
...
         super();
     }
 
     @httpGet('')
-    public async renderForm(): Promise<void> {
-        const session = await this.redisService.get(sessionKey);
+    public async renderForm(req: Request): Promise<void> {
+
+        const data = req.session
+            .chain(session => session.getExtraData())
```
Comment (n-r-b) [inline src/controllers/OtherReasonController.ts]: Could the middleware handle this?

File: src/controllers/OtherReasonController.ts [hunk 9]
```diff
@@ -5,33 +5,49 @@ import { SchemaValidator } from '../utils/validation/SchemaValidator';
...
+            .map(extraData => extraData[sessionKey]);
+
+        if (data.isJust()) {
```
Comment (n-r-b) [inline src/controllers/OtherReasonController.ts]: we could simplify this: `return await this.render(OK, data.isJust() ? data.__value: {});`

File: src/controllers/OtherReasonController.ts [hunk 10]
```diff
@@ -5,33 +5,49 @@ import { SchemaValidator } from '../utils/validation/SchemaValidator';
...
+import { EitherUtils, SessionMiddleware, SessionStore } from 'ch-node-session';
+import { AuthMiddleware } from '../middleware/AuthMiddleware';
 
 const sessionKey = 'session::other-reason';
 
-@controller(OTHER_REASON_PAGE_URI)
+@controller(OTHER_REASON_PAGE_URI, SessionMiddleware, AuthMiddleware)
 export class OtherReasonController extends BaseHttpController {
-    constructor(@inject(RedisService) private readonly redisService: RedisService) {
+    constructor(@inject(SessionStore) private readonly store: SessionStore) {
```
Comment (n-r-b) [inline src/controllers/OtherReasonController.ts]: can we call this `sessionStore`? as there is a `store` method on the SessionStore class

File: src/controllers/OtherReasonController.ts [hunk 11]
```diff
@@ -5,33 +5,49 @@ import { SchemaValidator } from '../utils/validation/SchemaValidator';
...
+        const body: OtherReason = req.body;
 
         const validationResult: ValidationResult = new SchemaValidator(schema).validate(body);
         const valid = validationResult.errors.length === 0;
 
         if (valid) {
-            await this.redisService.set(sessionKey, JSON.stringify(body))
+            req.session.map(async session => {
+                session.saveExtraData(sessionKey, body);
+                EitherUtils.wrapEitherFunction(Cookie.validateCookieString(req.cookies.__SID));
```
Comment (n-r-b) [inline src/controllers/OtherReasonController.ts]: Are we not validating the cookie in the session middleware?

File: src/controllers/OtherReasonController.ts (hunk 8 above)
Comment (franciscog-kainos) [reply src/controllers/OtherReasonController.ts]: What do you mean?

File: src/middleware/AuthMiddleware.ts [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+
+    public handler: RequestHandler = (req: Request, res: Response, next: NextFunction): void => {
+
+        const returnURL = this.popReturnURL();
+
+        if (req.cookies.__SID) {
+            const signInInfo = req.session.chain(session => session.getValue<ISignInInfo>(SessionKey.SignInInfo));
+            signInInfo.map(info => {
+                if (info[SignInInfoKeys.SignedIn] !== 1) {
+                    res.redirect(`${returnEnvVarible('REDIRECT_URL')}?returnTo=${returnURL}`);
```
[the remark this answers is not in the document]
Comment (geek-soft) [reply src/middleware/AuthMiddleware.ts]: Query param should be named `return_to` not `returnTo` otherwise redirect does not happen.

File: src/controllers/EntryController.ts [hunk 13]
```diff
@@ -1,11 +1,19 @@
...
 
+    constructor(@inject(AuthMiddleware) private readonly auth: AuthMiddleware) {
+        super();
+    }
+
     @httpGet('')
-    public redirectView(): void {
-        this.httpContext.response.redirect(PENALTY_DETAILS_PAGE_URI);
+    public async redirectView(): Promise<any> {
+        this.auth.setReturnURL(PENALTY_DETAILS_PAGE_URI);
```
[the remark this answers is not in the document]
Comment (franciscog-kainos) [reply src/controllers/EntryController.ts]: image

[the remark this answers is not in the document]
Comment (franciscog-kainos) [reply src/controllers/EntryController.ts]: The acceptance criteria knows best here

File: package.json [hunk 14]
```diff
@@ -43,6 +45,7 @@
     "@types/express": "^4.0.37",
     "@types/express-session": "^1.15.3",
     "@types/hapi__joi": "^16.0.10",
+    "@types/ioredis": "^4.14.7",
```
[the remark this answers is not in the document]
Comment (franciscog-kainos) [reply package.json]: After doing clean-install i get this when i run it
image

File: src/controllers/LandingController.ts [hunk 15]
```diff
@@ -1,17 +1,18 @@
...
+import { BaseAsyncHttpController } from './BaseAsyncHttpController';
 
 @controller(ROOT_URI)
-export class LandingController extends BaseHttpController {
+export class LandingController extends BaseAsyncHttpController {
 
     @httpGet('')
-    public renderView(req: Request, res: Response, next: NextFunction): void {
-        res.render('landing');
+    public async renderView(): Promise<string> {
```
[the remark this answers is not in the document]
Comment (franciscog-kainos) [reply src/controllers/LandingController.ts]: I'm testing these one by one to see which ones cause that dreadful `headers already sent to client `error

File: src/controllers/EntryController.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (geek-soft) [reply src/controllers/EntryController.ts]: As far as I understand the story (and other stories in either sprint or backlog) it should always take to penalty details screen after sign in. Please make it hardcoded. There is no need for such complexity. Let's keep it simple.

File: src/controllers/ConfirmationController.ts [hunk 16]
```diff
@@ -1,19 +1,19 @@
...
 
     @httpGet('')
-    public getConfirmationView(): void {
+    public getConfirmationView(req: Request): void {
 
-        const session = this.httpContext.request.session;
-        let companyNumber: string = '';
+        req.session.chain(_ => _.getExtraData())
+            .map(data => data.appeals.penaltyIdentifier.companyNumber)
+            .map(companyNumber => this.httpContext.response.render('confirmation', { companyNumber }))
```
Comment (geek-soft) [inline src/controllers/ConfirmationController.ts]: It would be ideal if we did not have to make two render calls and structure code in a way so that it reads session first and then it renders.

File: test/ApplicationFactory.ts [hunk 17]
```diff
@@ -1,10 +1,67 @@
...
+
+    const sessionStore = Substitute.for<SessionStore>();
+
+    if (session && cookie) {
+        sessionStore.load(cookie).returns(EitherUtils.wrapValue(session.data));
+    }
+
+    const realMiddleware = SessionMiddleware(config, sessionStore);
+    const sessionHandler = (req: Request, res: Response, next: NextFunction) => {
+        req.session = session ? Maybe.of(session) : Maybe.empty();
```
Comment (geek-soft) [inline test/ApplicationFactory.ts]: That seems to be not needed since real middleware will take care of it. Please remove.

File: test/controllers/OtherReasonController.test.ts [hunk 18]
```diff
@@ -51,11 +56,12 @@ describe('OtherReasonController', () => {
             const title = 'Some title';
             const description = 'Some description';
 
-            const app = createApplication(container => {
-                container.bind(RedisService).toConstantValue(createSubstituteOf<RedisService>(service => {
-                    service.set('session::other-reason', JSON.stringify({ title, description })).returns(Promise.resolve())
-                }))
-            });
+            const sessionStore = Substitute.for<SessionStore>();
```
Comment (geek-soft) [inline test/controllers/OtherReasonController.test.ts]: Please remove and put `{ title, description }` into session that is being passed to `createApp(session)` call.

File: test/controllers/OtherReasonDisclaimerController.test.ts [hunk 19]
```diff
@@ -1,29 +1,48 @@
...
 import { OK, MOVED_TEMPORARILY } from 'http-status-codes';
+import { CookieConfig, SessionStore, SessionMiddleware } from 'ch-node-session-handler';
+import { Redis } from 'ioredis';
+import { RequestHandler } from 'express';
+import { getEnvOrDefault } from '../../src/utils/EnvironmentUtils';
+
 
 describe('OtherReasonDisclaimerController', () => {
 
+    const app = createAppConfigurable(container => {
```
Comment (geek-soft) [inline test/controllers/OtherReasonDisclaimerController.test.ts]: Use the following combo here:
```
const session = createFakeSession([], config.cookieSecret, true);
const app = createApp(session);
```

:)

File: test/controllers/PenaltyDetailsController.test.ts [hunk 20]
```diff
@@ -1,55 +1,44 @@
...
-        it('should return 200 when trying to access page without a session', async () => {
-            await request(app).get(PENALTY_DETAILS_PAGE_URI)
-                .expect(response => {
-                    expect(response.status).to.be.equal(OK);
-                    expect(response.text).to.contain(pageHeading)
-                        .and.not.contain(errorSummaryHeading);
-                })
+        it('should return 302 when trying to access page without a session', async () => {
+            await request(createApp()).get(PENALTY_DETAILS_PAGE_URI)
+                .expect(MOVED_TEMPORARILY);
```
Comment (geek-soft) [inline test/controllers/PenaltyDetailsController.test.ts]: This might not be enough. Check where user agent was moved to.

Comment (geek-soft) [reply test/controllers/PenaltyDetailsController.test.ts]: Same sort of test should be applied to other controllers protected by auth middleware. Could be next PR if we decided to keep this PR small. If so this test could be moved to separate PR.

File: test/middleware/AuthMiddleware.test.ts [hunk 21]
```diff
@@ -0,0 +1,127 @@
...
+        let penaltyDetails: Record<string, any>;
+
+        before('sign in server should be running', async () => {
+
+            penaltyDetails = {
+                penaltyReference: 'A12345678',
+                companyNumber: 'SC123123'
+            };
+
+            await request(createApp())
```
Comment (geek-soft) [inline test/middleware/AuthMiddleware.test.ts]: I think there is no point having this precondition here. Application that is created and tested here is thrown away and not used in any of the tests executed afterwards.

File: test/middleware/AuthMiddleware.test.ts [hunk 22]
```diff
@@ -0,0 +1,127 @@
...
+        it('should not redirect the user to the sign in page if the user is signed in', async () => {
+
+            for (const page of protectedPages) {
+                await request(authedApp).get(page)
+                    .expect(200);
+            }
+
+        });
+
+        it('should allow the user to go to the penalty reference screen if authed', async () => {
```
Comment (geek-soft) [inline test/middleware/AuthMiddleware.test.ts]: This is essentially same as previous test case. I suggest you remove it.

File: test/middleware/AuthMiddleware.test.ts [hunk 23]
```diff
@@ -0,0 +1,127 @@
...
+
+            await request(authedApp)
+                .get(PENALTY_DETAILS_PAGE_URI)
+                .expect(200);
+
+        });
+        it('should continue to the next page without requiring auth if penalty details are valid', async () => {
```
Comment (geek-soft) [inline test/middleware/AuthMiddleware.test.ts]: There are few things I noticed about this test:
1) it targets penalty details page (integration) and it's behaviour while it should more focus on auth logic
2) it says 'if penalty details are valid' while validity of data does not matter at all since middleware is called before even controller gets the request and for that reason it is pointless

To fix it I suggest you change it to iterate through all the routes making POST request and assess that there is no redirect to sign in page.

Even better solution would be making this test a unit tests not integration tests but we have little time to change it.

File: test/middleware/AuthMiddleware.test.ts [hunk 24]
```diff
@@ -0,0 +1,127 @@
...
+const sig = generateSignature(id, config.cookieSecret);
+
+const session = createFakeSession(
+    [{
+        [SessionKey.Id]: id
+    }, {
+        [SessionKey.ClientSig]: sig
+    }], config.cookieSecret, true);
+
+const authedApp = createApp(session);
```
Comment (geek-soft) [inline test/middleware/AuthMiddleware.test.ts]: Since it is only used in first describe block I suggest you move it inside the block

File: test/middleware/AuthMiddleware.test.ts [hunk 25]
```diff
@@ -0,0 +1,127 @@
...
+            await request(authedApp)
+                .post(PENALTY_DETAILS_PAGE_URI)
+                .send(penaltyDetails)
+                .expect(_ => expect(_.header.location).to.include(OTHER_REASON_DISCLAIMER_PAGE_URI));
+
+        });
+    });
+    describe('Unauthed path', () => {
+
+        const appWithoutCookie = { label: 'App without Cookie', app: createApp(session, false) };
```
Comment (geek-soft) [inline test/middleware/AuthMiddleware.test.ts]: Label is not needed so please remove it. Variable name should be enough to express what it is.

Comment (geek-soft) [inline test/middleware/AuthMiddleware.test.ts]: Two things:
1) Auth middleware is not concerned with cookie.
2)` setCookie` in `const createApp = (session?: Session, setCookie: boolean = true)` is never used - likely because of 1)

I suggest you remove it due to above reasons.

File: test/middleware/AuthMiddleware.test.ts [hunk 26]
```diff
@@ -0,0 +1,127 @@
...
+        const unAuthedApp = { label: 'App without Cookie and Session', app: createApp() };
+
+        const unAuthedAppStateVariations = [appWithoutCookie, unAuthedApp];
+
+        before('sign in server should be running', async () => {
```
Comment (geek-soft) [inline test/middleware/AuthMiddleware.test.ts]: Same here, no value added, remove.

File: test/middleware/AuthMiddleware.test.ts [hunk 27]
```diff
@@ -0,0 +1,127 @@
...
+
+        });
+
+    });
+
+    describe('Auth Edge case', () => {
+
+        const appWithoutSessionObject = createApp(createFakeSession([], config.cookieSecret), true);
+
+        before('sign in server should be running', async () => {
```
Comment (geek-soft) [inline test/middleware/AuthMiddleware.test.ts]: Same here, no value added, remove.

File: test/middleware/AuthMiddleware.test.ts [hunk 28]
```diff
@@ -0,0 +1,127 @@
...
+
+            await request(createApp())
+                .get(PENALTY_DETAILS_PAGE_URI)
+                .catch(err => assert.fail(err));
+        });
+
+        it('should allow user to access protected pages if session object is somehow lost.', async () => {
```
Comment (geek-soft) [inline test/middleware/AuthMiddleware.test.ts]: Not sure what you trying to do here. What did you mean by 'is somehow lost'? A `createApp(createFakeSession([], config.cookieSecret), true)` call prepares session as other tests.

File: src/controllers/CheckYourAppealController.ts [hunk 29]
```diff
@@ -1,24 +1,36 @@
...
+                    reasons: data.appeals.reasons
+                };
+            });
 
-        if(session){
-            penaltyIdentifier = session.getExtraData('appeals').penaltyIdentifier;
-            reasons = session.getExtraData('appeals').reasons;
-            userProfile = session.getValue('signin_info').user_profile;
-        }
+        userProfile
```
Comment (geek-soft) [inline src/controllers/CheckYourAppealController.ts]: Mapping user profile and then getting appeals details does not make sense. Make just simple call to `this.httpContext.response.render` function passing both userProfile and appealsData. That mapping is absolutely necessary.

File: test/controllers/CheckYourAppealController.test.ts [hunk 30]
```diff
@@ -1,27 +1,25 @@
...
-const app = createApplication(container => {
-    container.bind(RedisService).toConstantValue(createSubstituteOf<RedisService>());
-});
+const config = getDefaultConfig();
 
 describe('CheckYourAppealController', () => {
     describe('GET request', () => {
-        it('should return 200 when trying to access the submission summary', async () => {
-
+        it('should return 200 when trying to access the submission summary with a session ', async () => {
```
Comment (geek-soft) [inline test/controllers/CheckYourAppealController.test.ts]: Not sure how much value that tests add. Not a concern of this PR but subject to removal in the future.

File: test/controllers/PenaltyDetailsController.test.ts (hunk 20 above)
Comment (akaJSaunders) [reply test/controllers/PenaltyDetailsController.test.ts]: We have removed this test and in another PR we can address more tests regarding auth middleware
